README.md
# Multi Select Form Field

A dropdown button Widget allowing one to choose multiple elements.

## Install

Add these line under your dependencies section:

```yaml
    multiselect_form_field:
        git:
            url: https://github.com/stevenosse/multi_select_form_field.git
            ref: master
```

## Demo

![alt text](./demo.gif "Demo gif")

### Features
- Regular widget
- Very simple to implement
- Can retrieve the list of selected elements
- Build custom list elements
- Can retrieve the list of unselected elements

### Example

```dart
MultiSelectFormField(
  key: _multiSelectKey,
  tagColor: Env.lightBlue,
  elementList: List.generate(
    15,
    (index) => MultiSelectFormFieldItem(
      labelStyle: TextStyle(fontWeight: FontWeight.w600),
      leading: CircleAvatar(),
      label: "Test $index",
      value: "test",
      isSelected: index.isEven,
    ),
  ),
),
```


## Retrieve more parameters

Give a Key to the widget, declared like : 
```dart
    final GlobalKey<MultiSelectFieldState> _multiSelectKey = GlobalKey();`
    
```

And then : 
```dart
    var selectedElements = _multiSelectKey.currentState.selectedElements; // Retrieve all the selected elements
    var unselectedElements = _multiSelectKey.currentState.unselectedElements; // Retrieve all the unselected elements
```
## Contribute
Every contributions are welcomed

CHANGELOG.md
## [0.0.1] - Initial release

* Initial functionnalities

## [0.0.2] - Bug fixes and plugin enhancement

* Replace the element list item from `Map` to `MultiSelectFormFieldItem`
* Made the selected element list scrollable (instead of letting it expand on bottom)
* Fixed the item selection bug
* Improved plugin customisation
* Other bug fixes and core improvements


lib/src/multi_select_form_field.dart
import 'package:flutter/material.dart';
import 'package:multiselect_form_field/src/multi_select_form_field_list_item.dart';
import 'package:multiselect_form_field/src/multi_select_form_field_item.dart';
import 'package:multiselect_form_field/src/multi_select_tag.dart';

class MultiSelectFormField extends StatefulWidget {
  final Key key;

  /// The list of elements to display
  ///
  final List<MultiSelectFormFieldItem> elementList;

  /// The color of the displayed tag color
  ///
  /// The default color is [Colors.lightBlue]
  final Color tagColor;

  /// The text to display when no element is selected
  ///
  /// The default text is ["No item selected yet"]
  final String emptyLabel;

  MultiSelectFormField({
    this.key,
    @required this.elementList,
    this.emptyLabel,
    this.tagColor,
  }) : super(key: key);

  @override
  MultiSelectFormFieldState createState() => MultiSelectFormFieldState();
}

class MultiSelectFormFieldState extends State<MultiSelectFormField> {
  /// Retrieve the list of selected elements
  get selectedElements =>
      widget.elementList.where((e) => e.isSelected == true).toList();

  /// Retrieve the list of unselected elements
  get unselectedElements =>
      widget.elementList.where((e) => e.isSelected == false).toList();

  @override
  Widget build(BuildContext context) {
    final Size size = MediaQuery.of(context).size;
    return Container(
      width: size.width,
      height: size.height * 0.3,
      child: Stack(
        children: <Widget>[
          Container(
            width: size.width,
            decoration: BoxDecoration(
              color: Colors.white,
              borderRadius: BorderRadius.circular(6.0),
              boxShadow: [
                BoxShadow(
                  blurRadius: 4.0,
                  color: Colors.black.withOpacity(0.15),
                ),
              ],
            ),
            child: Padding(
              padding: EdgeInsets.only(top: 55.0, left: 12.0, right: 12.0),
              child: ListView.builder(
                itemCount: widget.elementList.length,
                itemBuilder: (BuildContext context, int index) {
                  final element = widget.elementList[index];
                  return Material(
                    color: Colors.transparent,
                    child: InkWell(
                      onTap: () {
                        setState(() {
                          widget.elementList[index].isSelected = true;
                        });
                      },
                      child: Padding(
                        padding: const EdgeInsets.all(8.0),
                        child: Row(
                          children: <Widget>[
                            if (element.leading != null) ...[
                              element.leading,
                              SizedBox(width: 5)
                            ],
                            if (element.label != null)
                              Expanded(
                                child: Text(
                                  element.label,
                                  style: element.labelStyle.apply(
                                    color: element.isSelected
                                        ? Colors.black45
                                        : element.labelStyle.color,
                                  ),
                                ),
                              ),
                            if (element.trailing != null) ...[
                              SizedBox(width: 5),
                              element.trailing,
                            ],
                          ],
                        ),
                      ),
                    ),
                  );
                },
              ),
            ),
          ),
          _buildSelectedList(),
        ],
      ),
    );
  }

  _buildSelectedList() {
    final Size size = MediaQuery.of(context).size;
    return Container(
      width: size.width,
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.circular(4.0),
        boxShadow: [
          BoxShadow(
            blurRadius: 2.0,
            color: Colors.black.withOpacity(0.1),
          ),
        ],
      ),
      child: Padding(
        padding: EdgeInsets.all(4.0),
        child: SingleChildScrollView(
          scrollDirection: Axis.horizontal,
          child: Row(
            children: <Widget>[
              ..._buildSelectedElementList(),
            ],
          ),
        ),
      ),
    );
  }

  _buildSelectedElementList() {
    final elementList =
        widget.elementList.where((e) => e.isSelected == true).toList();
    if (elementList.isEmpty) return [_buildEmptyLabel()];
    return elementList
        .map(
          (e) => MultiSelectTag(
            label: e.label,
            tagColor: widget.tagColor,
            onRemove: () {
              widget.elementList.where((elt) => elt == e).first.isSelected =
                  false;
              setState(() {});
            },
          ),
        )
        .toList();
  }

  _buildEmptyLabel() {
    return Container(
      child: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Text(
          widget.emptyLabel ?? "No item selected yet",
          style: TextStyle(fontSize: 12.0, color: Colors.black38),
        ),
      ),
    );
  }
}


lib/src/multi_select_field_list_item.dart
import 'package:flutter/material.dart';

class MultiSelectFieldListItem extends StatelessWidget {
  final String label;
  final bool selected;
  final Widget leading;
  final VoidCallback onSelected;

  MultiSelectFieldListItem({
    @required this.label,
    this.selected = false,
    @required this.onSelected,
    @required this.leading,
  });

  @override
  Widget build(BuildContext context) {
    return Material(
      color: Colors.transparent,
      child: InkWell(
        onTap: () {
          if (selected == false) onSelected();
        },
        child: Padding(
          padding: const EdgeInsets.symmetric(vertical: 6.0),
          child: Row(
            children: <Widget>[
              if(leading != null) CircleAvatar(),
              SizedBox(width: 10),
              Text(
                "$label",
                style: TextStyle(
                  fontWeight: FontWeight.w600,
                  color: selected ? Colors.black38 : Colors.black,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

lib/src/multi_select_form_field_list_item.dart
import 'package:flutter/material.dart';

class MultiSelectFieldListItem extends StatelessWidget {
  /// The label of the item
  ///
  /// Takes a String as parameter
  final String label;

  /// Defines whether an item is selected or not
  ///
  /// The default value is [false]
  final bool selected;

  /// The leading widget
  ///
  /// If null, nothing will be displayed
  final Widget leading;

  /// The method called when an unselected item is tapped
  /// The element is of type [VoidCallback]
  final VoidCallback onSelected;

  /// The style to apply on each label element
  final TextStyle labelStyle;

  MultiSelectFieldListItem({
    @required this.label,
    this.selected = false,
    @required this.onSelected,
    this.leading,
    this.labelStyle,
  }) : assert(leading != null && label != null);

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 6.0),
      child: Row(
        children: <Widget>[
          if(leading != null) CircleAvatar(),
          SizedBox(width: 10),
          Text(
            "$label",
            style: labelStyle ?? TextStyle(
              fontWeight: FontWeight.w600,
              color: selected ? Colors.black38 : Colors.black,
            ),
          ),
        ],
      ),
    );
  }
}

lib/src/multi_select_tag.dart
import 'package:flutter/material.dart';

class MultiSelectTag extends StatelessWidget {
  final String label;
  final VoidCallback onRemove;
  final Color tagColor;

  MultiSelectTag({this.label, this.tagColor, @required this.onRemove});

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 4.0, vertical: 4.0),
      child: Container(
        decoration: BoxDecoration(
          color: tagColor ?? Colors.lightBlue,
          borderRadius: BorderRadius.circular(4.0),
        ),
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: Row(
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              Text(
                "$label",
                style: TextStyle(
                  fontSize: 12.0,
                  fontWeight: FontWeight.w600,
                ),
              ),
              SizedBox(width: 10),
              GestureDetector(
                onTap: onRemove,
                child: Icon(
                  Icons.close,
                  color: Colors.black26,
                  size: 15,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
